### ISSUE TITLE
crop_by_indices ignores shape_compensation when every box in the batch is the same (row output depends on the other rows)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4749]
## Summary

`crop_by_indices(..., size=..., shape_compensation="pad")` (any value other than `"resize"` pads) pads a box that is smaller than `size` — **unless every box in the batch is identical**, in which case the uniform-box fast path resizes it instead. The fast path never looks at `shape_compensation`. The same row with the same box therefore comes out padded or stretched depending on what the *other* rows of the batch contain, and a batch of one always takes the resize branch.

## Repro

```python
import torch
from kornia.geometry.transform import crop_by_indices

img = torch.arange(2 * 1 * 4 * 4, dtype=torch.float32).reshape(2, 1, 4, 4)
same = torch.tensor([[[0, 0], [1, 0], [1, 1], [0, 1]]] * 2)                        # both rows: the same 2x2 box
mixed = torch.tensor([[[0, 0], [1, 0], [1, 1], [0, 1]], [[1, 1], [2, 1], [2, 2], [1, 2]]])  # row 0 unchanged
for name, box in [("uniform", same), ("non-uniform", mixed)]:
    print(name, crop_by_indices(img, box, size=(3, 3), shape_compensation="pad")[0, 0].tolist())
print("B=1", crop_by_indices(img[:1], same[:1], size=(3, 3), shape_compensation="pad")[0, 0].tolist())
```

## Observed (main @ 8b03a8085, torch 2.14.0, CPU)

```
uniform     [[0.0, 0.5, 1.0], [2.0, 2.5, 3.0], [4.0, 4.5, 4.999999523162842]]   <- resized
non-uniform [[0.0, 1.0, 0.0], [4.0, 5.0, 0.0], [0.0, 0.0, 0.0]]                  <- padded
B=1         [[0.0, 0.5, 1.0], [2.0, 2.5, 3.0], [4.0, 4.5, 4.999999523162842]]   <- resized
```

Row 0 has the same box in both batched calls and still differs.

## Expected

The output of a row depends only on that row's box and on `shape_compensation`: the fast path should pad as well when `shape_compensation != "resize"` (or fall through to the general loop, which already honours it).

## Context

- `RandomMosaic` calls `crop_by_indices(..., shape_compensation="F.pad")`; #4679 works around exactly this by zero-padding the canvas first ("`crop_by_indices` resizes rather than pads when all boxes coincide"). A fix here should re-check that workaround and its pins.
- Found while reviewing #4741, whose `B > 0` guard on that fast path has a surviving `B > 1` mutant precisely because of this: for `B = 1` the two paths only differ under `shape_compensation="pad"`, and nothing pins that case. Not introduced by #4741.
- The same fast path also computes the slice and then falls through to the general loop whenever no resize is needed, so it only ever short-circuits the resize case.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
